Feature/1.7/caller includes pending time

lazan opened the pull request.

lazan committed:
service: removed intermediate service lookup reply

Before: If a requested service was busy, service-manager replied with an
intermediate busy reply, that hold the same information as the idle
reply later on. This was to enable caller to set "timers" and whatnot.
This has not been used for a while, so this patch got rid of it.

lazan committed:
common: more function style api for service::lookup

Made the service::lookup API more function style.
The `Lookup` type take care of the initial lookup, and then the user can
either 'consume' (move) the lookup and block for the reply, or optional non
blocking (non destructive) fetch reply can be used.

lazan committed:
service: service timeout includes pending lookup time

Now:
* If all instances of a 'local' service are busy, we keep the lookup
  in 'pending' and add timeout. When a timeout occur we also look in
  'pending'.
* A service call has a _timeout_ attribute. When a server gets invoked
  with a service call, it sets a deadline based on the _timeout_ and
  'now'. When a lookup for a service call is requested, the deadline is
  provided.
* service-manager uses the earliest deadline of the calculated deadline,
  based on the service timeout, or the provided deadline from caller.

issue #296

lazan committed:
gateway: handle service lookup timeout

This patch makes sure that inbound handle a timeout from service lookup.
Sends reply with 'TPETIME'

issue #296

lazan committed:
http: inbound handles lookup timeout

This patch makes sure that http-inbound reacts correctly on service
lookup timeout.

issue #296

AlbsAgren commented on middleware/common/include/common/message/service.h:
```diff
@@ -307,7 +320,6 @@ namespace casual
                   {
                      regular,
                      no_reply,
-                     no_busy_intermediate,
```
Feels nice to be rid of this, always found it a bit confusing.

AlbsAgren commented on middleware/common/include/common/service/lookup.h:
```diff
@@ -18,95 +18,56 @@ namespace casual
-            Lookup& operator = ( Lookup&&) noexcept;
+      struct Lookup
+      {
+         //Lookup() noexcept;
```
🗑️

AlbsAgren commented on middleware/common/source/service/lookup.cpp:
```diff
@@ -12,145 +12,111 @@ namespace casual
+      Lookup::~Lookup()
+      {
+         // we have to discard if we're still pending with the service manager
+         if( ! m_correlation)
```
If m_correlation is empty I dont think it makes sense to call local::discard since that function has a guard that returns if the correlation is empty? Should it be `if( m_correlation)`?

lazan replied:
Of course!

AlbsAgren reviewed (commented):
Doing it this way makes a lot of sense to me

lazan committed:
common: execution id should be inherited by children

Before: When execution id was reset, we did not update/set the
corresponding (internal) environment variable -> every child process
from a unittest (domains and so on) had it's own execution id. This
makes it unnecessary hard to filter out a particular unittest from a log
file with hundreds of unittests.

Now: We set the (internal) environment virable.

Note: We don't actually set to the system environment variables, we set
to our own _prioritized_ set of variables, that child processes gets
during spawn.

lazan committed:
transaction: manager should wait for all spawned RM:s before on-line

Before: transcation-manager did not wait for all spawned RM:s to be
ready before connect to domain-manager. This gave some unpredictable
results in unittests, where the distribution of interactions with RM:s
was not deterministic.

Now: transaction-manager sends connect request to DM when all RM:s are
'ready' (they could have died also).

I don't think this would have been an issue at all for real usage of
casual, but in our unittests where we spawn hundreds of domains it
matters.

lazan committed:
refactoring: service timeout -> service call deadline

The naming _call deadline_ makes it clearer that the deadline is not
necessary bound to whatever timeout the service has.
A service call gets the earliest deadline (timepoint) of the deadline it
already has (if any) and the timeout of the requested service.

Also: Cleaned up common/message/service.h. There's different needs to
what a _service_ represent in different context.
Before: we had a lot of inheritance (code reuse) which made it
unnecessary hard to maintain the different needs over time. Unnecessary
attributes got sucked in, which made it harder to understand, besides
the fact that we sent unnecessary bytes.
Now: Every _service type_ is it's own bespoke type.

issue #296

lazan commented:
Merged manually
